test/mpi/autogen.sh

```diff
@@ -0,0 +1,178 @@
+        timelimit=`echo $line | sed -E 's|(.+):(.*):(.+):(.*):(.+)|\4|'`
+        procs=`echo $line | sed -E 's|(.+):(.*):(.+):(.*):(.+)|\5|'`
+        other_macros=""
+    else
+        # the line is a commet
+        pathname=`echo $line | sed -E 's|(.+):(.*):(.+):(.*):(.+)|\1|'`
+        ln=$((ln+1)) # increment line number
+        continue
+    fi
+
+    dir=`dirname $pathname`
+    source=`echo $(basename $pathname) | sed -E "s|(.+)\.c[xx]*|\1|"`
+    ext=`echo $(basename $pathname) | sed -E "s|.+\.(c[xx]*)|\1|"`
+
+    # check whether .dtp and .in files already exist in dir ...
+    exists=0
+    for i in $dirs
+    do
+        if [ $i = $dir ] ; then
+            exists=1
+        fi
+    done
+
+    # ... if not create them
+    if [ $exists = 0 ] ; then
+        printf "" > ${dir}/Makefile.dtp
+        printf "" > ${dir}/testlist.dtp
+        printf "" > ${dir}/testlist.in
+        dirs+="$dir "
+    fi
+
+    # prepare user defined macros
+    for macro in $macros
+    do
+        other_macros+="-D$macro "
+    done
+
+    sendcounts=$counts
+
+    printf "basictypetest.txt:${ln}: Generate tests for: ${source}.${ext} ... "
+    tn=1 # basic MPI datatype number in basictypelist.txt
+    for type in $types
+    do
+        recvcounts=$sendcounts # reset recv counts to send counts
+
+        scn=1 # send count number in basictypetest.txt:ln
+        for sendcount in $sendcounts
+        do
+            if [ $dir = "pt2pt" ] ; then # only send/recv comm can use types from different pools
+                # do combination of different send recv count where recv count >= send count
+                rcn=$scn
+                for recvcount in $recvcounts
+                do
+                    echo "noinst_PROGRAMS += ${source}_${tn}_${scn}_${rcn}_${ln}" >> ${dir}/Makefile.dtp
+                    echo "${source}_${tn}_${scn}_${rcn}_${ln}_CPPFLAGS = -DDTP_BTYPE_TMPL=$type -DDTP_BTYPE_SENDCOUNT_TMPL=$sendcount -DDTP_BTYPE_RECVCOUNT_TMPL=$recvcount $other_macros \$(AM_CPPFLAGS)" >> ${dir}/Makefile.dtp
+                    echo "${source}_${tn}_${scn}_${rcn}_${ln}_SOURCES = ${source}.${ext}" >> ${dir}/Makefile.dtp
+                    echo "${source}_${tn}_${scn}_${rcn}_${ln} $procs $timelimit" >> ${dir}/testlist.dtp
+                    rcn=$((rcn+1)) # increment recv count number
+                done
+                recvcounts=`echo $recvcounts | sed -e "s|$sendcount||"` # update recv counts
+            else
+                echo "noinst_PROGRAMS += ${source}_${tn}_${scn}_${ln}" >> ${dir}/Makefile.dtp
+                echo "${source}_${tn}_${scn}_${ln}_CPPFLAGS = -DDTP_BTYPE_TMPL=$type -DDTP_BTYPE_COUNT_TMPL=$sendcount $other_macros \$(AM_CPPFLAGS)" >> ${dir}/Makefile.dtp
+                echo "${source}_${tn}_${scn}_${ln}_SOURCES = ${source}.${ext}" >> ${dir}/Makefile.dtp
+                echo "${source}_${tn}_${scn}_${ln} $procs $timelimit" >> ${dir}/testlist.dtp
+            fi
```

Comment:
My initial concern with adding macros, counts and types to the name is that this can end up being quite large. Currently the name contains references that can help reconstruct the test configuration starting from configurations files. This is more compact. If you prefer I can add a description of how test names are generated in the README.

Reply:
This has been fixed to print type signature information in the executable name. However this is not done for the struct types defined in `structtypetest.txt` because it's not practical for structs with large number of elements.